README.md
# QuickSkip for Kodi 

<img src="icon.png" alt="QuickSkip logo"/>

## Introduction

This add-on for Kodi (former XBMC) allows you to fast skip commercials in TV show recordings with a few button presses. The main idea is to skip the recording forward in 180 or 60 second steps as long as you see that the TV show is running again. After that, you can skip backward or forward to find the exact position where the commercials end and the TV show continues. All you need to do is to press the `up` key during video playback and then press `right` or `left` keys to find the position where the commercials end. Similar functionality was provided in Topfield PVR set-top boxes as an TAP add-on named [Skip It](http://www.saunalahti.fi/tsalomak/skipit.html).

Typical usage scenario is this:
1. You are playing a recorded TV show and commercials are starting. You want to skip to place where the commercials have ended and the TV show continues.
1. Press `up` key to show the add-on dialog.
1. Press `right` to skip 180 seconds forward.
1. You see that commercials are still running. Press `right` key again to skip 180 seconds forward.
1. You see that commercials are not running anymore so you skipped too far. Press `left` key to skip 180 / 2 = 90 seconds backwards.
1. You see that commercials are still not running so you have to skip a little bit more backwards. Press `left` key again to skip 90 / 2 = 45 seconds backwards.
1. You see that commercials are now running so you have to skip a little bit more forward. Press `right` key to skip 45 / 2 = 22 seconds forward.
1. You see that the commercial break is just ending and the TV show continues. Don't press anything and the add-on dialog will disappear in a few seconds.

Usage might sound complicated at first but it is very powerful, intuitive and fast way to skip commercials. You can skip commercials in just a few seconds by pressing `right` and `left` keys to the direction where the commercial break is ending and TV show continues.

## Supported Kodi versions

This add-on has been tested to work well at least in LibreELEC 7.0.3 with Kodi version 16.1. There seems to be some problems in 
LibreELEC 8.0.1 with Kodi version 17.1. The add-on seems to work most of the time with Kodi 17.1 but the dialog does not always auto close for some reason.
Also skipping seems to be a little bit slower with Kodi 17.1 than in 16.1.

## Installation to Kodi

1. Download this repository as [zip file](https://github.com/mvestola/plugin.video.quick.skip/archive/master.zip).
1. Copy the zip file to the computer you are running Kodi.
1. Install the plugin from the zip file using the Kodi add-on manager
1. Define the key that will launch the add-on. You have two options to do this:
   1. Use [Keymap editor add-on](http://kodi.wiki/view/Add-on:Keymap_Editor) to configure the key. Select `Programs...Keymap Editor...Edit...Fullscreen Video...Add-ons...Launch QuickSkip` and Input the `up` key. Finally save Keymap Editor settings.
   1. Create user defined keymap file to your computer running Kodi to location `~/.kodi/userdata/keymaps/quick-skip-keymap.xml` with contents same as in file [/resources/data/quick-skip-keymap.xml](/resources/data/quick-skip-keymap.xml). Reboot Kodi.

## Usage in Kodi

1. Open some video file and play it in fullscreen mode
1. Click `up` key from your keyboard or remote control: a dialog window should appear to the top left corner
1. Use `left` and `right` arrow keys from your keyboard or remote control to skip to direction you want
1. You can press `OK` key to change the time to skip from 180 sec -> 60 sec -> 10 sec.
1. Close the dialog with `Esc` from your keyboard or `back` from your remote control. Or just wait for a few seconds and the dialog will automatically close.

You can change some settings for the add-on in the add-on settings page in Kodi. Values that can be changed are:
* Time in seconds to auto close the dialog
* First, second and third skip values in seconds which default to 180 / 60 / 10 seconds. These are the skip seconds that change when you press the `OK` key when add-on dialog is visible.

## Screenshots

<img src="resources/screenshots/screenshot-01.jpg" alt="Add-on view with logo" />
<img src="resources/screenshots/screenshot-02.jpg" alt="Add-on settings" />
<img src="resources/screenshots/screenshot-03.jpg" alt="Add-on dialog when playing video" />

## Changelog

### 1.0.0 (2017-04-15)
* First stable release

## Contact

You can contact the developer by email: quick.skip@mvnet.fi

For reporting bugs, use the GitHub issue tracker.


changelog.txt
CHANGELOG

1.0.0
- First stable release


addon.py
import xbmcaddon
import xbmcgui
import xbmc
from threading import Timer

# To launch this script, add the file "/resources/data/quick-skip-keymap.xml" to your kodi user defined keymaps: ~/.kodi/userdata/keymaps/

addon = xbmcaddon.Addon()

# See key codes from https://github.com/xbmc/xbmc/blob/master/xbmc/input/Key.h
ACTION_MOVE_LEFT =  1
ACTION_MOVE_RIGHT = 2
ACTION_MOVE_UP = 3
ACTION_PAGE_UP =  5
ACTION_SELECT_ITEM = 7
ACTION_PREVIOUS_MENU = 10
ACTION_NAV_BACK = 92
ACTION_CHANNEL_UP = 184

xbfont_left = 0x00000000
xbfont_right = 0x00000001
xbfont_center_x = 0x00000002
xbfont_center_y = 0x00000004
xbfont_truncated = 0x00000008

SKIP_LEFT = "left"
SKIP_RIGHT = "right"

AUTO_CLOSE_TIMEOUT_SECONDS = int(addon.getSetting("auto_close_timeout"))
FIRST_SKIP_SECONDS = int(addon.getSetting("first_skip_seconds"))
SECOND_SKIP_SECONDS = int(addon.getSetting("second_skip_seconds"))
THIRD_SKIP_SECONDS = int(addon.getSetting("third_skip_seconds"))

i18n_jump = addon.getLocalizedString(32201)
i18n_place = addon.getLocalizedString(32202)
i18n_sec = addon.getLocalizedString(32203)

timeout = None

class QuickSkipDialog(xbmcgui.WindowDialog):

    def __init__(self):
        self.skipSeconds = float(FIRST_SKIP_SECONDS)
        self.cumulativeSkipSeconds = 0.0
        self.startTime = None
        self.previousDirection = None
        self.directionChanged = False

        self.addControl(xbmcgui.ControlImage(10, 10, 200, 90, addon.getAddonInfo('path')+'/resources/media/background.png'))

        self.controlLabel1 = xbmcgui.ControlLabel(x=20, y=20, width=180, height=30, label="", alignment=xbfont_center_y|xbfont_center_x)
        self.addControl(self.controlLabel1)
        self.controlLabel1.setLabel(i18n_jump + ": " + str(FIRST_SKIP_SECONDS) + " " + i18n_sec)

        self.controlLabel2 = xbmcgui.ControlLabel(x=20, y=60, width=180, height=30, label="", alignment=xbfont_center_y|xbfont_center_x)
        self.addControl(self.controlLabel2)
        self.controlLabel2.setLabel(i18n_place + ": +00:00:00")

    def restartTimeoutTimer(self):
        global timeout
        timeout.cancel()
        timeout = Timer(AUTO_CLOSE_TIMEOUT_SECONDS, self.close)
        timeout.start()

    def closeDialogByUserInput(self):
        global timeout
        timeout.cancel()
        del timeout
        self.close()

    def onAction(self, action):
        action_id = action.getId()
        if action_id == ACTION_SELECT_ITEM or action_id == ACTION_MOVE_UP or action_id == ACTION_PAGE_UP or action_id == ACTION_CHANNEL_UP:
            if not self.directionChanged:
                if int(self.skipSeconds) == FIRST_SKIP_SECONDS:
                    self.restartTimeoutTimer()
                    self.skipSeconds = float(SECOND_SKIP_SECONDS)
                    self.controlLabel1.setLabel(i18n_jump + ": " + str(SECOND_SKIP_SECONDS) + " " + i18n_sec)
                elif int(self.skipSeconds) == SECOND_SKIP_SECONDS:
                    self.restartTimeoutTimer()
                    self.skipSeconds = float(THIRD_SKIP_SECONDS)
                    self.controlLabel1.setLabel(i18n_jump + ": " + str(THIRD_SKIP_SECONDS) + " " + i18n_sec)
                else:
                    self.closeDialogByUserInput()
            else:
                self.closeDialogByUserInput()
        elif action_id == ACTION_MOVE_LEFT:
            self.handleSkip(SKIP_LEFT)
        elif action_id == ACTION_MOVE_RIGHT:
            self.handleSkip(SKIP_RIGHT)
        else:
            self.closeDialogByUserInput()


    def handleSkip(self, direction):
        self.restartTimeoutTimer()

        if self.previousDirection is not None and self.previousDirection != direction:
            self.directionChanged = True

        if not self.directionChanged and self.startTime is None:
            self.startTime = xbmc.Player().getTime()

        if self.directionChanged:
            self.skipSeconds = self.skipSeconds / 2

        if abs(self.skipSeconds) < 2.0:
            self.skipSeconds = 2.0

        if direction == SKIP_LEFT:
            self.skipSeconds = -abs(self.skipSeconds)
        else:
            self.skipSeconds = abs(self.skipSeconds)

        self.cumulativeSkipSeconds = self.cumulativeSkipSeconds + self.skipSeconds
        self.previousDirection = direction

        m, s = divmod(abs(self.cumulativeSkipSeconds), 60)
        h, m = divmod(m, 60)
        formatted_place = "%02d:%02d:%02d" % (h, m, s)

        if self.cumulativeSkipSeconds < 0.0:
            formatted_place = "-" + formatted_place
        else:
            formatted_place = "+" + formatted_place

        self.controlLabel1.setLabel(i18n_jump + ": " + str(int(self.skipSeconds)) + " " + i18n_sec)
        self.controlLabel2.setLabel(i18n_place + ": " + formatted_place)

        place_to_seek = self.startTime + self.cumulativeSkipSeconds

        if place_to_seek < 0:
            place_to_seek = 0

        xbmc.Player().seekTime(place_to_seek)


if xbmc.Player().isPlayingVideo():
    dialog = QuickSkipDialog()
    timeout = Timer(AUTO_CLOSE_TIMEOUT_SECONDS, dialog.close)
    timeout.start()
    dialog.doModal()
    del dialog
else:
    xbmcgui.Dialog().ok(addon.getLocalizedString(32301), addon.getLocalizedString(32302))
